File: ts/day_6/main.test.ts
import {
  assertEquals,
  assertNotEquals,
} from "https://deno.land/std@0.117.0/testing/asserts.ts";
import {
  bench,
  BenchmarkTimer,
  runBenchmarks,
} from "https://deno.land/std@0.117.0/testing/bench.ts";
import * as day from "./main.ts";

const part1Solution = 360761;
const part2Solution = 1632779838045;

Deno.test("part 1 real", () => {
  const input = day.readData();

  assertEquals(day.part1(input), part1Solution);
});

Deno.test("part 2 real", () => {
  const input = day.readData();

  assertEquals(day.part2(input), part2Solution);
});

Deno.test("parse contents", () => {
  const input = "3,4,3,1,2";

  assertEquals(day.parseContents(input), [0, 1, 1, 2, 1, 0, 0, 0, 0]);
});

Deno.test("part 1 example", () => {
  const input: day.fish = [0, 1, 1, 2, 1, 0, 0, 0, 0];

  assertEquals(day.part1(input), 5934);
});

Deno.test("part 2 example", () => {
  const input: day.fish = [0, 1, 1, 2, 1, 0, 0, 0, 0];

  assertEquals(day.part2(input), 26984457539);
});

bench({
  name: "read data",
  runs: 5000,
  func(b: BenchmarkTimer): void {
    b.start();
    const input = day.readData();
    assertNotEquals(input, []);
    b.stop();
  },
});

bench({
  name: "part 1",
  runs: 5000,
  func(b: BenchmarkTimer): void {
    const input = day.readData();
    b.start();
    assertEquals(day.part1(input), part1Solution);
    day.part1(input);
    b.stop();
  },
});

bench({
  name: "part 2",
  runs: 5000,
  func(b: BenchmarkTimer): void {
    const input = day.readData();
    b.start();
    assertEquals(day.part2(input), part2Solution);
    b.stop();
  },
});

if (Deno.args.length > 0 && Deno.args[0] == "--bench") {
  runBenchmarks();
}

File: go/day_6/main_test.go
package main

import (
	"reflect"
	"testing"
)

func TestReadData(t *testing.T) {
	data := readData("../../data")
	if reflect.DeepEqual(data, Input{}) {
		t.FailNow()
	}
}

func TestParseContents(t *testing.T) {
	tests := map[string]struct {
		data     string
		expected Input
	}{
		"example": {
			data:     "3,4,3,1,2",
			expected: Input{0, 1, 1, 2, 1, 0, 0, 0, 0},
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			actual := parseContents(test.data)

			if !reflect.DeepEqual(actual, test.expected) {
				t.Fatalf("Expected: %v\nActual: %v", test.expected, actual)
			}
		})
	}
}

func TestPart1(t *testing.T) {
	tests := map[string]struct {
		data     Input
		expected string
	}{
		"example": {
			data:     Input{0, 1, 1, 2, 1, 0, 0, 0, 0},
			expected: "5934",
		},
		"actual": {
			data:     readData("../../data"),
			expected: part1Solution,
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			actual := test.data.Part1()

			if actual != test.expected {
				t.Fatalf("Expected: %v\nActual: %v", test.expected, actual)
			}
		})
	}
}

func TestPart2(t *testing.T) {
	tests := map[string]struct {
		data     Input
		expected string
	}{
		"example": {
			data:     Input{0, 1, 1, 2, 1, 0, 0, 0, 0},
			expected: "26984457539",
		},
		"actual": {
			data:     readData("../../data"),
			expected: part2Solution,
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			actual := test.data.Part2()

			if actual != test.expected {
				t.Fatalf("Expected: %v\nActual: %v", test.expected, actual)
			}
		})
	}
}

func TestFishAfterDays(t *testing.T) {
	tests := map[string]struct {
		data     Input
		days     int
		expected int
	}{
		"18 days example": {
			data:     Input{0, 1, 1, 2, 1, 0, 0, 0, 0},
			days:     18,
			expected: 26,
		},
		"80 days example": {
			data:     Input{0, 1, 1, 2, 1, 0, 0, 0, 0},
			days:     80,
			expected: 5934,
		},
		"256 days example": {
			data:     Input{0, 1, 1, 2, 1, 0, 0, 0, 0},
			days:     256,
			expected: 26984457539,
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			actual := test.data.fishAfterDays(test.days)

			if actual != test.expected {
				t.Fatalf("Expected: %v\nActual: %v", test.expected, actual)
			}
		})
	}
}

func BenchmarkReadData(b *testing.B) {
	for i := 0; i < b.N; i++ {
		if reflect.DeepEqual(readData("../../data"), Input{}) {
			b.FailNow()
		}
	}
}

func BenchmarkPart1(b *testing.B) {
	data := readData("../../data")
	for i := 0; i < b.N; i++ {
		if data.Part1() != part1Solution {
			b.FailNow()
		}
	}
}

func BenchmarkPart2(b *testing.B) {
	data := readData("../../data")
	for i := 0; i < b.N; i++ {
		if data.Part2() != part2Solution {
			b.FailNow()
		}
	}
}
